src/merkle/store.rs

```diff
@@ -0,0 +1,375 @@
+
+        for (index_value, mut node, path) in paths {
+            let mut index = NodeIndex::new(self.nodes.len() as u8, index_value);
+
+            for sibling in path {
+                let (left, right) = match index.is_value_odd() {
+                    true => (sibling, node),
+                    false => (node, sibling),
+                };
+                let parent = Rpo256::merge(&[left.into(), right.into()]).into();
+                self.nodes.insert(
+                    parent,
+                    Node {
+                        left: left.into(),
+                        right: right.into(),
+                    },
+                );
+
+                index.move_up();
+                node = parent.word;
+            }
+        }
+
+        // Returns the parent of the last paths (assumes all paths have the same parent) or empty
+        Ok(node)
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn get_node(&self, root: Word, index: NodeIndex) -> Result<Word, MerkleError> {
+        let mut hash: WordWithOrd = root.into();
+        let directions = index.directions();
+        for dir in directions {
+            let node = self
+                .nodes
+                .get(&hash)
+                .ok_or(MerkleError::NodeNotInStorage(hash.into(), index))?;
+            hash = match dir {
+                Direction::Left => node.left,
+                Direction::Right => node.right,
+            };
+        }
+
+        Ok(hash.into())
+    }
+
+    pub fn get_path(&self, root: Word, index: NodeIndex) -> Result<MerklePath, MerkleError> {
+        let mut hash: WordWithOrd = root.into();
+        let mut path = Vec::new();
+        for dir in index.directions() {
+            let node = self
+                .nodes
+                .get(&hash)
+                .ok_or(MerkleError::NodeNotInStorage(hash.into(), index))?;
+            hash = match dir {
+                Direction::Left => {
+                    path.push(node.right.into());
+                    node.left
+                }
+                Direction::Right => {
+                    path.push(node.left.into());
+                    node.right
+                }
+            }
+        }
+
+        Ok(MerklePath::new(path))
+    }
+
+    // DATA MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn set_node(
+        &mut self,
+        root: Word,
+        index: NodeIndex,
+        value: Word,
+    ) -> Result<(), MerkleError> {
+        let path = self.get_path(root, index)?;
+
+        let paths = [(index.value(), value, path)];
+        self.add_merkle_paths(paths).map(|_| ())
+    }
+
+    pub fn merge_roots(&mut self, root1: Word, root2: Word) -> Result<Word, MerkleError> {
+        let parent: Word = Rpo256::merge(&[root1.into(), root2.into()]).into();
```

Comment:
Should we check here if both `root1` and `root2` are in the store, and if not return an error?

Reply:
Added. Note that we can't check if the values root1 and root2 are actually roots, since we don't nave the `NodeIndex` anywhere. It would probably make this much more costly to track and update these values, here for example we would have to walk both trees and change all the indexes.

Reply:
Yeah, I think just checking that these values are in the store is fine.